modules/3d/perf/perf_rendering.cpp

```diff
@@ -0,0 +1,215 @@
+            if (v >= 0 && v < 5)
+            {
+                *os << svals[v];
+            }
+            else
+            {
+                *os << "UNKNOWN";
+            }
+        }
+        static ::testing::internal::ParamGenerator<OutputsEnum> all()
+        {
+            return ::testing::Values(OutputsEnum(vals[0]),
+                                     OutputsEnum(vals[1]),
+                                     OutputsEnum(vals[2]));
+        }
+
+    private:
+        Outputs val;
+    };
+
+    const std::array<Outputs, 3> OutputsEnum::vals{ Outputs::DepthOnly,
+                                                    Outputs::ColorOnly,
+                                                    Outputs::DepthColor
+                                                  };
+    const std::array<std::string, 3> OutputsEnum::svals{ std::string("DepthOnly"),
+                                                         std::string("ColorOnly"),
+                                                         std::string("DepthColor")
+                                                       };
+
+    static inline void PrintTo(const OutputsEnum &t, std::ostream *os) { t.PrintTo(os); }
+}
+
+
+// resolution, shading type, outputs needed
+typedef perf::TestBaseWithParam<std::tuple<std::tuple<int, int>, ShadingTypeEnum, OutputsEnum>> RenderingTest;
+
+PERF_TEST_P(RenderingTest, rasterizeTriangles, ::testing::Combine(
+    ::testing::Values(std::make_tuple(1920, 1080), std::make_tuple(1024, 768), std::make_tuple(640, 480)),
+    ShadingTypeEnum::all(),
+    OutputsEnum::all()
+    ))
+{
+    auto t = GetParam();
+    auto wh = std::get<0>(t);
+    int width = std::get<0>(wh);
+    int height = std::get<1>(wh);
+    auto shadingType = std::get<1>(t);
+    auto outputs = std::get<2>(t);
+
+    string objectPath = findDataFile("rendering/spot.obj");
+
+    Vec3f position = Vec3f(20.0, 60.0, -40.0);
+    Vec3f lookat   = Vec3f( 0.0,  0.0,   0.0);
+    Vec3f upVector = Vec3f( 0.0,  1.0,   0.0);
+
+    double fovy = 45.0;
+
+    std::vector<Vec3f> vertices;
+    std::vector<Vec3i> indices;
+    std::vector<Vec3f> colors;
+    std::vector<vector<int>> indvec;
+    loadMesh(objectPath, vertices, colors, indvec);
+    for (const auto &vec : indvec)
+    {
+        indices.push_back({vec[0], vec[1], vec[2]});
+    }
+
+    if (shadingType == ShadingType::White)
+    {
+        colors.clear();
+    }
+    else
+    {
+        for (auto &color : colors)
+        {
+            color = Vec3f(abs(color[0]), abs(color[1]), abs(color[2])) * 255.0f;
+        }
+    }
+
+    double zNear = 0.1, zFar = 50;
+
+    Mat cameraMatrix = Mat(makeCamMatrix_TODO_rewrite_it_later(position, lookat, upVector, fovy, zNear, zFar));
+
+    Mat depth_buf, color_buf;
+    while (next())
+    {
+        // Prefilled to measure pure rendering time w/o allocation and clear
+        depth_buf = Mat(height, width, CV_32F, zFar);
+        color_buf = Mat(height, width, CV_32FC3, Scalar(0.0, 0.0, 0.0));
```

Comment:
Just 0 or Scalar::all(0) can be used as the last parameter. Even better to use cv::Mat::zeros for it.